api/v1beta1/azuremachine_validation.go

```diff
@@ -293,3 +297,37 @@ func validateCachingType(cachingType string, fieldPath *field.Path, managedDisk
 	allErrs = append(allErrs, field.Invalid(cachingTypeChildPath, cachingType, fmt.Sprintf("allowed values are %v", compute.PossibleCachingTypesValues())))
 	return allErrs
 }
+
+// ValidateDiagnostics validates the Diagnostic spec.
+func ValidateDiagnostics(diagnostics Diagnostics, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if diagnostics.Boot != nil {
+		if diagnostics.Boot.StorageAccountType == UserManagedDiagnosticsStorage {
+			if diagnostics.Boot.UserManaged == nil {
+				allErrs = append(allErrs, field.Required(fieldPath.Child("UserManaged"),
+					fmt.Sprintf("userManaged must be specified when storageAccountType is '%s'", UserManagedDiagnosticsStorage)))
+			} else if diagnostics.Boot.UserManaged.StorageAccountURI == "" {
+				allErrs = append(allErrs, field.Required(fieldPath.Child("StorageAccountURI"),
+					fmt.Sprintf("storageURI cannot be empty when storageAccountType is '%s'", UserManagedDiagnosticsStorage)))
+			}
+		}
+
+		possibleStorageAccountTypeValues := []BootDiagnosticsStorageAccountType{
+			DisabledDiagnosticsStorage,
+			ManagedDiagnosticsStorage,
+			UserManagedDiagnosticsStorage,
+		}
+
+		for _, possibleValue := range possibleStorageAccountTypeValues {
+			if diagnostics.Boot.StorageAccountType == possibleValue {
+				return allErrs
+			}
+		}
+
+		allErrs = append(allErrs, field.Invalid(fieldPath.Child("StorageAccountType"), diagnostics.Boot.StorageAccountType,
+			fmt.Sprintf("allowed values are %v", possibleStorageAccountTypeValues)))
```

Comment:
this is code is unreachable, AFAICT it's already validated via the openAPI schema Enum annotation